### DISCUSSION TITLE
How do I use English special characters while in normal mode

### DISCUSSION THREAD

**Discussion by RobberFox** [Wansmer/langmapper.nvim#43] [Q&A]
When I'm using russian, most special characters are shifted or don't exist. Thus I want to achieve the next thing:

#### While using russian keyboard layout in non-insert mode:
1. Map `ёЁ` to `` `~`` so I can use marks and uppercase action.
2. Map Shift+number row from `"№;%:?` to `@#$%^&` so that I have access to `^$%` horizontal motions, `@` macros, etc.
3. Map every other special character, i.e. from `бю.жэхъ` and `БЮ,ЖЭХЪ` to `,./;'[]` and `<>?:"{}`, so that I have access to `f` finding special characters at the expense of not being able to find those 6 russian letters.

Can I achieve that, and how?

###### P.S. Not sure if this discussion should be in Russian or English

**Comment by Wansmer** [comment]
You should configure `vim.opt.langmap` for this (see the readme for an example). The `langmapper` is only for keybindings, not for normal movements.
You can also read this article to understand in depth how the langmapper works: https://habr.com/ru/articles/726400/

> P.S. Not sure if this discussion should be in Russian or English

English, because not only Russian use `langmapper` and have similar question. English more universal in this case

**Comment by RobberFox** [comment]
Yeah, I only managed to map `` `~ `` *(they weren't working for some reason)*:
```lua
vim.opt.langmap:append(vim.fn.join({"ёЁ;`~"}, ","))
```
But I don't know what to do with the points 2. and 3. since langmap from one special character to another special character is logically wrong.
It seems to be outside of the scope of `langmap` this is why I decided to ask here.

Also, how do you deal with not having `$^%` keys when using the russian layout?

And in case none of that is possible, would you recommend using something like `https://github.com/lyokha/vim-xkbswitch` alongside this plugin?

**Comment by Wansmer** [comment]
> point 2

This is solved by adding the necessary character mappings to `default_layout` and `layouts.ru.layout`, and setting the `get_current_layout_id` function for your OS. Check the README.

> point 3

I'm not sure what you're trying to achieve. The `f` and `t` are looking for the characters you enter. It makes sense that special symbols are in different places on different layouts. You can make your own layout change by `f` and `t`, but that is beyond the scope of this plugin.

> And in case none of that is possible, would you recommend using something like https://github.com/lyokha/vim-xkbswitch alongside this plugin?

I have never used this plugin and can't recommend it. If it solves your problems, try it

**Comment by RobberFox** [comment]
Almost works
```lua
default_layout = [[ABCDEFGHIJKLMNOPQRSTUVWXYZ<>?:"{}~@#$^&abcdefghijklmnopqrstuvwxyz,./;'[]`]],
layouts = {
	ru = {
		id = "ru",
		layout = [[ФИСВУАПРШОЛДЬТЩЗЙКЫЕГМЦЧНЯБЮ,ЖЭХЪË"№;:?фисвуапршолдьтщзйкыегмцчнябю.жэхъё]],
	},
},
os = {
	Linux = {
		get_current_layout_id = function()
			local cmd = "xkb-switch"
			if vim.fn.executable(cmd) then
				local output = vim.split(vim.trim(vim.fn.system(cmd)), "\n")

				return output[#output]
			end
		end,
	},
},
```
But now while using the russian layout I have overriden the same keys as the english `,.:;"` keys (i.e. while in russian, pressing the english `,`, `.`, `;`, `:`, `"` gives `?`, `/`, `$`, `^`, `@` respectively) How can I fix this?

P.S. And I guess the step 3 isn't something that I necessarily need. I can just alter my `ru` layout system-wide where I have 3rd Alt-gr layer with english symbols.

**Comment by RobberFox** [comment]
Is this a bug, or I am doing something wrong? If it is, should I make a bug report?

**Comment by Wansmer** [comment]
> But now while using the russian layout I have overriden the same keys as the english ,.:;@ keys (i.e. while in russian, pressing ,, ., ;, :, " gives ?, /, $, ^, @ respectively) How can I fix this?

If I understand you correctly, you are describing the expected behaviour, because when you are in the Russian layout and press `.` you get the `/` functionality, and when you press `ю` you get the `.` functionality. This is done intentionally so that your muscle memory does not break down depending on the current input method. 
Also make sure there is no recursion in your mappings (`layout`, `default_layout`, `vim.opt.langmap`).

**Comment by RobberFox** [comment]
What I meant is that in the overriden keys I lose access to the default-us-layout behaviour, for example If I have.
```lua
default_layout = [[ABCDEFGHIJKLMNOPQRSTUVWXYZ<>?:"{}~abcdefghijklmnopqrstuvwxyz,./;'[]`]],
layouts = {
	ru = {
		...
		layout = [[ФИСВУАПРШОЛДЬТЩЗЙКЫЕГМЦЧНЯБЮ,ЖЭХЪËфисвуапршолдьтщзйкыегмцчнябю.жэхъё]],
		...
	},
}
```
Then while in the russian layout those 3 keys are overriden like such:
![image](https://github.com/user-attachments/assets/3ebf61e6-fbf3-49de-b805-d6a1c22551af)

So I don't have access to comma and dot repeat keys. Which I, obviously, want to access because otherwise it breaks down my muscle memory depending on the current input method.

**Reply by Wansmer** [reply]
I checked these comparisons with myself. All three keys work correctly in both layouts. 

Maybe your `get_current_layout_id` is not working correctly, but I can't check that because I'm not using Linux and xkb-switch. Make sure that the current layout names are returned and that these are the names specified as id in the plugin settings.

**Reply by RobberFox** [reply]
I checked and output of "xkb-switch" matches the `id = "ru"`:
![image](https://github.com/user-attachments/assets/abc3f70a-ea89-4307-9552-9195a4716b88)
And that function seems to be working because I get at least *some* behaviour that's not present in the default config.
But testing this as a minimal config I still have the same issue of the comma and dot keys being overriden:
```lua
local function escape(str)
	local escape_chars = [[;,."|\]]
	return vim.fn.escape(str, escape_chars)
end

local en = [[`qwertyuiop[]asdfghjkl;'zxcvbnm,.]]
local ru = [[ёйцукенгшщзхъфывапролджэячсмитьбю]]
local en_shift = [[~QWERTYUIOP{}ASDFGHJKL:"ZXCVBNM<>]]
local ru_shift = [[ËЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЯЧСМИТЬБЮ]]

vim.opt.langmap = vim.fn.join({
	escape(ru_shift) .. ";" .. escape(en_shift),
	escape(ru) .. ";" .. escape(en),
}, ",")

local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not (vim.uv or vim.loop).fs_stat(lazypath) then
	local lazyrepo = "https://github.com/folke/lazy.nvim.git"
	local out = vim.fn.system({ "git", "clone", "--filter=blob:none", "--branch=stable", lazyrepo, lazypath })
	if vim.v.shell_error ~= 0 then
		vim.api.nvim_echo({
			{ "Failed to clone lazy.nvim:\n", "ErrorMsg" },
			{ out, "WarningMsg" },
			{ "\nPress any key to exit..." },
		}, true, {})
		vim.fn.getchar()
		os.exit(1)
	end
end

vim.opt.rtp:prepend(lazypath)


require("lazy").setup({
	spec = {
		{
			"Wansmer/langmapper.nvim",
			lazy = false,
			priority = 1,

			config = function()
				require("langmapper").setup({
					map_all_ctrl = true,
					ctrl_map_modes = { "n", "o", "i", "c", "t", "v" },
					hack_keymap = true,
					disable_hack_modes = { "i" },
					automapping_modes = { "n", "v", "x", "s" },

					default_layout = [[ABCDEFGHIJKLMNOPQRSTUVWXYZ<>?:"{}~abcdefghijklmnopqrstuvwxyz,./;'[]`]],

					use_layouts = {},
					custom_desc = nil,

					layouts = {
						ru = {
							id = "ru",
							layout = [[ФИСВУАПРШОЛДЬТЩЗЙКЫЕГМЦЧНЯБЮ,ЖЭХЪËфисвуапршолдьтщзйкыегмцчнябю.жэхъё]],
						},
					},

					os = {
						Linux = {
							get_current_layout_id = function()
								local cmd = "xkb-switch"

								return vim.notify(vim.trim(vim.fn.system(cmd)))
							end,
						},
					},

				})
			end,
		},
	},
})

require("langmapper").automapping({ global = true, buffer = true })
```
I don't really know to do.

**Reply by Wansmer** [reply]
`vim.notify` don't return string: 

```lua
Linux = {
  get_current_layout_id = function()
    local cmd = "xkb-switch"

    return vim.notify(vim.trim(vim.fn.system(cmd)))
  end,
}
```

**Reply by RobberFox** [reply]
Sorry, I accidentally sent the wrong code, it's without notify

**Reply by Wansmer** [reply]
I checked it again with minimal config (but using `get_current_layout_id` for Darwin) - all work correctly:

https://github.com/user-attachments/assets/3844ce71-ab96-4e6c-9f65-1e22f5db0cc0


Unfortunately, I don't know why it doesn't work in your case. Try looking for the configs for `langmapper`, which also uses Linux and `xkb-switch`.

**Comment by RobberFox** [comment]
Well, I can do this:
```lua
default_layout = [[/?@#$^&]],
layouts = {
	ru = {
		id = "ru",
		layout = [[.,"№;:?]],
	},
},
```
The special characters now work, but any Ctrl keybinds don't

**Comment by RobberFox** [comment]
So, it seems like (at least on my system) if I remove the last `init.lua` line:
```lua
require("langmapper").automapping({ global = true, buffer = true })
```
Then this functionality works as intended.

P.S. This is a personal log at this point, I don't want to waste others time, so no problem if you don't respond

**Comment by RobberFox** [comment]
Now (without the automapping) having:
```lua
default_layout = [[@#$^&ABCDEFGHIJKLMNOPQRSTUVWXYZ<>:"{}~?abcdefghijklmnopqrstuvwxyz,.;'[]`/]],
layouts.ru.layout = [["№;:?ФИСВУАПРШОЛДЬТЩЗЙКЫЕГМЦЧНЯБЮЖЭХЪË,фисвуапршолдьтщзйкыегмцчнябюжэхъё.]],
```
I can access symbols from the russian layout, but it works only in the normal mode. In the visual mode, for example `$^` don't work and instead those are russian `;:`
